pkg/network/kuryr_bootstrap.go

```diff
@@ -0,0 +1,844 @@
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get port list")
+	}
+	portList, err := ports.ExtractPorts(page)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to extract ports list")
+	}
+	if len(portList) > 1 {
+		return "", errors.Errorf("found multiple ports matching name %s, tag %s, cannot proceed", name, tag)
+	} else if len(portList) == 1 {
+		return portList[0].ID, nil
+	} else {
+		opts := ports.CreateOpts{
+			Name: name,
+			NetworkID: netId,
+			FixedIPs: []ports.IP{{SubnetID: subnetId, IPAddress: ip}},
+		}
+		portObj, err := ports.Create(client, opts).Extract()
+		if err != nil {
+			return "", errors.Wrap(err, "failed to create port")
+		}
+
+		_, err = tagResource(client, "ports", portObj.ID, tag)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to tag created port")
+		}
+
+		return portObj.ID, nil
+	}
+}
+
+// Looks for a OpenStack security group by name and tag. Fails if SG is not found
+// or multiple SG's match.
+func findOpenStackSgId(client *gophercloud.ServiceClient, name, tag string) (string, error) {
+	page, err := groups.List(client, groups.ListOpts{Name: name, Tags: tag}).AllPages()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get SG list")
+	}
+	sgs, err := groups.ExtractGroups(page)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to extract SG list")
+	}
+
+	if len(sgs) == 1 {
+		return sgs[0].ID, nil
+	} else if len(sgs) == 0 {
+		return "", errors.New("SG not found")
+	} else {
+		return "", errors.New("multiple matching SGs")
+	}
+}
+
+// Looks for a OpenStack security group by name and tag. If it does not exist
+// creates it. Will fail if multiple SG's are matching all criteria.
+func ensureOpenStackSg(client *gophercloud.ServiceClient, name, tag string) (string, error) {
+	page, err := groups.List(client, groups.ListOpts{Name: name, Tags: tag}).AllPages()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get SG list")
+	}
+	sgs, err := groups.ExtractGroups(page)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to extract SG list")
+	}
+	if len(sgs) > 1 {
+		return "", errors.Errorf("found multiple SG matching name %s, tag %s, cannot proceed", name, tag)
+	} else if len(sgs) == 1 {
+		return sgs[0].ID, nil
+	} else {
+		opts := groups.CreateOpts{
+			Name: name,
+		}
+		sg, err := groups.Create(client, opts).Extract()
+		if err != nil {
+			return "", errors.Wrap(err, "failed to create SG")
+		}
+
+		_, err = tagResource(client, "security-groups", sg.ID, tag)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to tag created SG")
+		}
+
+		return sg.ID, nil
+	}
+}
+
+// Tries to create an OpenStack security group rule on sgId SG. Ignores an
+// error if such rule already exists.
+func ensureOpenStackSgRule(client *gophercloud.ServiceClient, sgId, remoteSgId string) error {
+	opts := rules.CreateOpts{
+		SecGroupID: sgId,
+		EtherType: rules.EtherType4,
+		Direction: rules.DirIngress,
+		RemoteGroupID: remoteSgId,
```

Comment:
use subnetpool cird instead, for scalability reasons

Reply:
On a second look, it this even used? or just the following method ensureOpenStackSgRuleCIDR? In that case, perhaps it can be removed

Reply:
Yeah, I forgot to replace ensureOpenStackSgRule with ensureOpenStackSgRuleCIDR. Fixed on my local copy, will update this PR once openshift/api commit is merged and I can finally make this pass all the tests.